src/main/scala/vectorpipe/osm/internal/functions/package.scala

```diff
@@ -0,0 +1,676 @@
+      case Seq(h, t @ _ *) =>
+        val x = h.getX(h.size - 1)
+        val y = h.getY(h.size - 1)
+
+        formRings(t.find(line => x == line.getX(0) && y == line.getY(0)) match {
+          case Some(next) =>
+            h.append(new PartialCoordinateSequence(next, 1)) +: t.filterNot(line => isEqual(line, next))
+          case None =>
+            t.find(line => x == line.getX(line.size - 1) && y == line.getY(line.size - 1)) match {
+              case Some(next) =>
+                h.append(
+                  new PartialCoordinateSequence(
+                    new ReversedCoordinateSequence(next), 1
+                  )
+                ) +: t.filterNot(line => isEqual(line, next))
+              case None => throw new AssemblyException("Unable to connect segments.")
+            }
+        }, rings)
+    }
+  }
+
+  implicit val geometryFactory: GeometryFactory = new geom.GeometryFactory()
+
+  private def formRings(segments: GenTraversable[LineString])(implicit geometryFactory: GeometryFactory): GenTraversable[Polygon] =
+    formRings(segments.map(_.getCoordinateSequence).map(s => new VirtualCoordinateSequence(Seq(s))))
+      .map(geometryFactory.createPolygon)
+
+  // join segments together
+  @tailrec
+  private def connectSegments(
+    segments: GenTraversable[VirtualCoordinateSequence],
+    lines: Seq[CoordinateSequence] = Vector.empty[CoordinateSequence]
+  ): GenTraversable[CoordinateSequence] = {
+    segments match {
+      case Nil =>
+        lines
+      case Seq(h, t @ _ *) =>
+        val x = h.getX(h.size - 1)
+        val y = h.getY(h.size - 1)
+
+        t.find(line => x == line.getX(0) && y == line.getY(0)) match {
+          case Some(next) =>
+            connectSegments(
+              h.append(new PartialCoordinateSequence(next, 1)) +: t.filterNot(line => isEqual(line, next)),
+              lines
+            )
+          case None =>
+            t.find(line => x == line.getX(line.size - 1) && y == line.getY(line.size - 1)) match {
+              case Some(next) =>
+                connectSegments(
+                  h.append(
+                    new PartialCoordinateSequence(new ReversedCoordinateSequence(next), 1)
+                  ) +: t.filterNot(line => isEqual(line, next)),
+                  lines
+                )
+              case None => connectSegments(t, lines :+ h)
+            }
+        }
+    }
+  }
+
+  private def connectSegments(segments: GenTraversable[LineString])(implicit geometryFactory: GeometryFactory): GenTraversable[LineString] =
+    connectSegments(segments.map(_.getCoordinateSequence).map(s => new VirtualCoordinateSequence(Seq(s))))
+      .map(geometryFactory.createLineString)
+
+  private def getHoles(poly: Polygon): List[CoordinateSequence] = {
+    for (i <- 0 until poly.getNumInteriorRing) yield
+      poly.getInteriorRingN(i).getCoordinateSequence
+  }.toList
+
+  private def dissolveRings(rings: GenTraversable[Polygon]): (Seq[Polygon], Seq[Polygon]) = {
```

Comment:
This no longer dissolves. (That's what the unary `union` in `MultiPolygon(rings.toArray).union.asMultiPolygon` did.)

Reply:
That's something I meant to comment on. Do you know how to make the JTS Geometries dissolve?

Reply:
Not off the top of my head.

Reply:
Okay, I'll ask around and see if anyone knows

Reply:
@mojodna So I think this does perform the dissolve; as it still dissolves all of the rings within the polygon.

Reply:
Actually, this is something that the unit tests in `MultiPolygonRelationReconstructionSpec` will check (there's at least 1 relation made up up multiple ways w/o rings), so perhaps it'd be best to add that and use it to confirm.

Reply:
I did a little poking; once you have a JTS `MultiPolygon`, call `.union()` on it and you should get either a `Polygon` or `MultiPolygon` back.

Something like this should work:

```scala
val mp = factory.createMultiPolygon(rings.toArray).union
val ps = for (i <- 0 until mp.getNumGeometries) yield mp.getGeometryN(i).asInstanceOf[Polygon]
(ps.map(_.getExteriorRing).map(x => factory.createPolygon(x.getCoordinateSequence)),
 ps.flatMap(getHoles).map(factory.createPolygon))
```